MBS-12715, MBS-13526: ArtistCreditEditor refactor

mwiencek opened the pull request:
This is based on top of https://github.com/metabrainz/musicbrainz-server/pull/3111 since I use the new library for updating React state here.

# Problem

The current ArtistCreditEditor has a number of issues:

 * It has no Flow types.
 * It's using old autocomplete/bubble components which also have no Flow types, and have been replaced by other components (`Autocomplete2`, `ButtonPopover`).
 * It's using ugly hacks to keep the bubble in sync with the editor component, instead of `createPortal`.
 * It's written as a class-based component, which React no longer recommends.
 * It maintains its own internal React state which must be synchronized with other view models from the inside, instead of being stateless and fully controlled by parent components.
 * It logs of bunch of warnings due to incorrect usage of `flushSync`.

In case the motivation is not clear: the above issues currently make it difficult to use the ArtistCreditEditor in future edit forms, like the React conversion of the release editor, and the "alternative tracklists" editor.

# Solution

This basically refactors the entire component and its sub-components, and updates a lot of glue code that keeps it in sync with Knockout observables in the release, recording, and release group edit forms.

# Testing

Just manual testing so far in the release, recording, and release group edit forms.
 * Copying/pasting artist credits.
 * Using "guess feat. artists" in the recording edit form.
 * Moving between tracks in the release editor, including using the enter key.

reosarevok commented on root/static/scripts/edit/components/ArtistCreditBubble.js:
```diff
@@ -1,172 +1,304 @@
+    const handleEnter = () => {
+      if (
+        isTrack &&
+        // $FlowIgnore[prop-missing]
```
When is this missing? When there's no next track? If so, cannot we just check whether the prop exists at all?

mwiencek replied:
Well, `entity` is an `ArtistCreditableT` which may or may not be a track in the release editor. Checking `isTrack` just doesn't tell Flow that it's a track. I'll try to improve it.

reosarevok reviewed (commented):
Lots to go yet, few comments for now.

mwiencek commented on root/static/scripts/common/components/Dialog.js:
```diff
@@ -164,6 +164,16 @@ const Dialog = ({
+        if (!(event.target instanceof HTMLButtonElement)) {
+          event.preventDefault();
+          event.target.dispatchEvent(
+            new Event('mb-dialog-enter', {bubbles: true}),
```
I listen for it in `ArtistCreditBubble` to determine whether to go to the next track or close the bubble. It was much easier than passing an event handler through a bunch of nested components.

mwiencek commented on root/static/scripts/common/components/EntityLink.js:
```diff
@@ -187,8 +194,10 @@ const EntityLink = ({
 }: EntityLinkProps):
 $ReadOnlyArray<Expand2ReactOutput> | Expand2ReactOutput | null => {
   const hasCustomContent = nonEmpty(passedContent);
+  // $FlowIgnore[sketchy-null-mixed]
```
Well, I think these are the graceful ways... :)
Checking if the properties are expected to exist would either require `hasOwnProperty` checks (which are unnecessary, and wouldn't convince Flow anyway, or tell you if they are expected), or `entityType` refinements (which might convince Flow, but are also completely unnecessary).

reosarevok replied:
Ok, ok, I guess there's no sensible way to just make Flow trust us. I just feel these sort of ignores are asking for bugs down the line, but probably not too bad here.

mwiencek commented on root/static/scripts/edit/components/ArtistCreditBubble.js:
```diff
@@ -1,172 +1,304 @@
+      style={FLOAT_LEFT}
+      type="button"
+    >
+      {l('Copy Credits')}
```
I changed "Copy credits," "Paste credits," "Add artist credit," and "Remove artist credit" (the last one was only a title attribute). Let me know if I missed anything.

mwiencek commented on root/static/scripts/edit/components/ArtistCreditBubble.js:
```diff
@@ -1,172 +1,304 @@
+    >
+      {l('Done')}
+    </button>
+    {isTrack ? (
```
I don't think it would be crazily difficult, but would rather not stuff more issues into this PR. :) Perhaps in a follow-up one?

reosarevok replied:
Sure, fair enough. Added MBS-13396

mwiencek commented on root/static/scripts/edit/components/ArtistCreditBubble.js:
```diff
@@ -1,172 +1,304 @@
+  <tr>
+    <td colSpan="3" style={PADDING_BOTTOM_1EM}>
+      {exp.l(
+        `Use the following fields to enter collaborations. See the
```
Same as above. I'm just re-using the existing string and would rather keep this PR focused on the refactoring for now!

Aerozol replied:
'Use the following fields to enter artist name variations and multiple artist collaborations.'
Is the first thing to come to mind.

mwiencek replied:
Merged the new string into the `ArtistCreditDocumentation` component.

mwiencek commented:
@brainzbot test this please

reosarevok commented:
Tested this locally, seems to work fine - except for the automatic feat. standardization, but that's good since that hasn't been our guideline for years now so that code should be removed anyway. Do check if other join phrase standardizations are still wanted though that also fail?

reosarevok commented on root/static/scripts/edit/components/ArtistCreditNameEditor.js:
```diff
@@ -78,62 +100,81 @@ class ArtistCreditNameEditor extends React.Component {
+  return (
+    <tr>
+      {artistCreditName.removed ? (
+        <td colSpan="3" style={{textAlign: 'right'}}>
```
Should this style be a class?

reosarevok reviewed (commented):
Some comments, checking this up to the start of the `release-editor` folder.

mwiencek commented on root/static/scripts/common/utility/entityHref.js:
```diff
@@ -87,7 +88,7 @@ function entityHref(
 
     default:
-      if (entityProps.mbid && entity.gid) {
+      if (entityProps.mbid && nonEmpty(entity.gid)) {
```
It would always see it as non-empty, yes. But the `nonEmpty` call was mainly added to work around a sketchy-null error from Flow. If the unwrapped `gid` is empty, `entityHref` will return a nonsense result either way (because `id` will just remain an empty string).

mwiencek replied:
I'm guessing that's why I originally added https://github.com/metabrainz/musicbrainz-server/pull/3113#discussion_r1408145768 though.

mwiencek commented on root/static/scripts/edit/components/ArtistCreditBubble.js:
```diff
@@ -1,172 +1,302 @@
-  }
-  const pressedEsc = event.which === 27;
-  const pressedEnter = event.which === 13;
+const FLOAT_LEFT = {float: 'left'};
+const FLOAT_RIGHT = {float: 'right'};
+const WIDTH_40PC = {width: '40%'};
+const PADDING_BOTTOM_1EM = {paddingBottom: '1em'};
```
No, I think the previous code inlined these and I just moved them into constants (which is slightly better than before). I can look at turning them into CSS.

mwiencek commented:
@brainzbot test this please

mwiencek commented:
@brainzbot test this please

mwiencek commented:
@brainzbot test this please

reosarevok commented:
There's a conflict here, but in general maybe we should put this on test?

mwiencek commented:
> There's a conflict here, but in general maybe we should put this on test?

Sure, I've rebased and put it on the test server now. :)

reosarevok commented:
Few comments from testing on test.mb:

* Is it intentional that the popover takes less space than before, and that "Done" is no longer right-aligned? (I assume the change from grey to white is intentional since it shows the buttons a lot more clearly)
* Is it intentional that the entered credit is kept when blanking the artist field with the keyboard, but only if it differs from the default fro the artist being deleted? (seems useful but it differs from prod where it seemingly always gets removed when blanking the field, so making sure it's a feature and not a bug)
* Do we store at all the [removed] entries in state? Would it be possible to un-remove if I click by mistake? (before it was annoying to misclick but you sighed and moved on... but now it reminds you it *knows* there was something, so it kinda feels you should be able to get it back!)
* If I have at least two empty artist fields, I can end up with multiple Recent items dropdowns at once. It seems to mostly be reproduceable by clicking one empty field, clicking somewhere else, then clicking the other empty field, see:
![Peek 2024-02-19 16-02](https://github.com/metabrainz/musicbrainz-server/assets/1069224/11b97b75-41c5-4843-9d33-75bd143305b2)
* Probably the weirdest nitpick ever, but in case it might show some other underlying issue: if I blank a join phrase between artists 1 and 2, then I add artist 3, the blanked phrase remains blanked. If I *paste* an AC with a blank join phrase between artists 1 and 2, then I add artist 3, the blanked phrase turns into `, `.
* More nitpicking: Does the positioning of the search icon look weird to you in the release tracklist? It does to me, it kinda feels like it should be either also green so it looks like the artist field, or exactly in the center of the white square?
![Screenshot from 2024-02-19 16-22-43](https://github.com/metabrainz/musicbrainz-server/assets/1069224/7bb504e3-3b61-4d45-a39b-4b996525df94)
* Semi-relatedly, should we grey out that search icon or have a hover text or something so that it is clearer that it will work for track 1 here but not track 2? (this is not a new issue but for some reason - probably because the icon is bigger now - it feels more annoying than before). It still shows "Search" on hover, maybe it could at least show "Cannot search for multiple artists" or something?
![Screenshot from 2024-02-19 16-24-48](https://github.com/metabrainz/musicbrainz-server/assets/1069224/6023c40b-d28c-4982-9892-5809a16b35ee)

yvanzo commented:
Also when removing a name, the line stays with `[removed]` and the red cross button, whereas in production the line just disappears.

yvanzo commented:
> ArtistCreditEditor refactor (including MBS-12715)

It still is the main change here so the usual `MBS-12715: Refactor artist credit editor` would do.

reosarevok commented:
The `[removed]` thing is expected, it's apparently an accessibility thing (that's why the relationship editor already does it).

yvanzo commented:
> The `[removed]` thing is expected, it's apparently an accessibility thing (that's why the relationship editor already does it).

Ok, I guess that is a first step towards your above point 3 about un-removing (as the relationship editor already does it).

mwiencek commented:
> Is it intentional that the popover takes less space than before

Do you mean *more* space? I had increased the max-width, and it's clearly bigger for me.

> and that "Done" is no longer right-aligned?

Fixed

> I assume the change from grey to white is intentional since it shows the buttons a lot more clearly

I changed it back to gray, since nobody complained about it before

> Is it intentional that the entered credit is kept when blanking the artist field with the keyboard, but only if it differs from the default fro the artist being deleted? (seems useful but it differs from prod where it seemingly always gets removed when blanking the field, so making sure it's a feature and not a bug)

I'm not sure if this was intentional, but if you prefer the behavior we could keep it.

> Do we store at all the [removed] entries in state? Would it be possible to un-remove if I click by mistake? (before it was annoying to misclick but you sighed and moved on... but now it reminds you it knows there was something, so it kinda feels you should be able to get it back!)

You can undo them now

> If I have at least two empty artist fields, I can end up with multiple Recent items dropdowns at once. It seems to mostly be reproduceable by clicking one empty field, clicking somewhere else, then clicking the other empty field, see: [...]

Fixed

> Probably the weirdest nitpick ever, but in case it might show some other underlying issue: if I blank a join phrase between artists 1 and 2, then I add artist 3, the blanked phrase remains blanked. If I paste an AC with a blank join phrase between artists 1 and 2, then I add artist 3, the blanked phrase turns into , .

Fixed

> More nitpicking: Does the positioning of the search icon look weird to you in the release tracklist? It does to me, it kinda feels like it should be either also green so it looks like the artist field, or exactly in the center of the white square?

I tweaked the positioning and it turns green now. (However, this looks kinda weird for inputs that still have a border -- unlike the ones in the tracklist.)

> Semi-relatedly, should we grey out that search icon or have a hover text or something so that it is clearer that it will work for track 1 here but not track 2? (this is not a new issue but for some reason - probably because the icon is bigger now - it feels more annoying than before). It still shows "Search" on hover, maybe it could at least show "Cannot search for multiple artists" or something?

This sounds like a good idea, but I'm not actually sure how to gray out the icon (and it's an Autocomplete2 issue, not an ArtistCreditEditor issue -- I'd prefer to work on other improvements later).

yvanzo reviewed (approved):
It looks good to me. I didn’t test the latest changes but you obviously did.

Just a couple of minor tidying-up points:

* “Removal undoing” at least seems to be worth a separate ticket.
* The PR title should start to with the addressed ticket(s) [as usual](https://github.com/metabrainz/guidelines/blob/master/GitHub.md), even though some more refactoring is done as well.
